Param net

PhMueller opened the pull request.

PhMueller committed:
[skip travis] Rebasing

PhMueller committed:
[skip travis] Initial Commit ParamNet Benchmark

PhMueller committed:
[skip travis] Initial Commit ParamNet Benchmark

PhMueller committed:
[skip travis] ParamNet Split Benchmarks in Time-Optimized and "Step-Optimized"

PhMueller committed:
[skip travis] Update Tests + Add Container support

PhMueller committed:
[skip travis] Merge remote-tracking branch 'origin/ParamNet' into ParamNet

PhMueller committed:
[skip travis] Update dependencies

PhMueller committed:
Final Commit Paramnet

PhMueller committed:
[skip travis] Redo some changes

KEggensperger commented on hpobench/benchmarks/surrogates/paramnet_benchmark.py:
```diff
@@ -0,0 +1,380 @@
+    def convert_config_to_array(configuration: Dict) -> np.ndarray:
+        cfg_array = np.zeros(8)
+        cfg_array[0] = 10 ** configuration['initial_lr_log10']
+        cfg_array[1] = round(2 ** configuration['batch_size_log2'])  # todo: round
```
This todo is outdated?

KEggensperger replied:
Maybe make a comment saying that in contrast to previous versions we now round

KEggensperger commented on hpobench/benchmarks/surrogates/paramnet_benchmark.py:
```diff
@@ -0,0 +1,380 @@
+
+        seed = seed if seed is not None else np.random.randint(1, 100000)
+        cs = CS.ConfigurationSpace(seed=seed)
+        # TODO: in origin benchmark: all uniform float and default is the mid value.
```
Then why did you change the default value for this version?

PhMueller replied:
good question. I've changed the values to the original default values but wrote them explicitly in the constructor so that it is clear what the default values are.

KEggensperger commented on hpobench/benchmarks/surrogates/paramnet_benchmark.py:
```diff
@@ -0,0 +1,380 @@
+
+        Fidelities
+        ----------
+        step: float - [1, 50]
```
integer

KEggensperger commented on hpobench/benchmarks/surrogates/paramnet_benchmark.py:
```diff
@@ -0,0 +1,380 @@
+        # If we can't afford a single epoch, return TODO.
+        if (costs / self.n_epochs) > fidelity['budget']:
+            # TODO: Return random performance here instead
+            y = 1
```
No, 1 is fine. Are all of these classification problems and we minimize error in [0, 1]?

KEggensperger commented on hpobench/benchmarks/surrogates/paramnet_benchmark.py:
```diff
@@ -0,0 +1,380 @@
+        learning_curves_cost = np.linspace(costs / self.n_epochs, costs, self.n_epochs)
+
+        if fidelity['budget'] < costs:
+            idx = np.where(learning_curves_cost < fidelity['budget'])[0][-1]
```
It is very unlikely, but learning_curve_cost==budget can happen. Shouldn't this be <= then?

KEggensperger commented on hpobench/benchmarks/surrogates/paramnet_benchmark.py:
```diff
@@ -0,0 +1,380 @@
+            lc = lc[:idx]
+
+        else:
+            # If the budget is larger than the actual runtime, we extrapolate the learning curve
```
This has been done before, could you please add a line to the meta-info where you took the this from (I assume hpolib1.5)

KEggensperger commented on hpobench/benchmarks/surrogates/paramnet_benchmark.py:
```diff
@@ -0,0 +1,380 @@
+        -------
+        ConfigSpace.ConfigurationSpace
+        """
+        budgets = {  # .   (min, max)-budget (seconds) for the different data sets
```
Please add a link to the HPBandster Code to know where these numbers come from

KEggensperger replied:
Ah, okay, if this is also described in the Paper (as in meta-info), then we don't need a link here

KEggensperger commented on hpobench/container/recipes/surrogates/Singularity.ParamnetBenchmark:
```diff
@@ -0,0 +1,21 @@
+    cd /home \
+    && git clone https://github.com/automl/HPOBench.git \
+    && cd HPOBench \
+    && pip install --upgrade .[paramnet] \
```
Why --upgrade? We don't do this for other benchmarks?

PhMueller replied:
This is because we need an older scikit-learn version. Saving and restoring an TreeRegressor is only allowed until 0.25 I guess. There is a warning about it. Therefore, I put the new scikit-learn version in the requirements and to make sure it is installed, I've added the --upgrade parameter. 
Without it, pip is already happy if there is it already installed.

KEggensperger replied:
Okay, so this is fine then.

KEggensperger commented on hpobench/util/data_manager.py:
```diff
@@ -537,7 +539,6 @@ def __init__(self, data_path: Union[str, Path]):
                             lock_path=f'{hpobench.config_file.cache_dir}/lock_nasbench_101_data', delay=0.5)
     def _download(self):
         from tqdm import tqdm
```
I am fine with adding tqdm as a dependency, but then it should be part of requirements.txt and not only required for paramnet?

PhMueller replied:
hm, the thing is, we only use it for the paramnet benchmark (more precisely in the paramnet dataloader). The Hpobench package does not rely on it. But if you prefer it more in the general requirements, that's no problem at all :-)

KEggensperger replied:
But this is not the paramnet dataloader, this is NASBench_101. The paramnet dataloader does not make use of tqdm. Okay, but then, since it is only one dataloader, I guess it's best to move the dependency to the respective benchmark requirements.

PhMueller replied:
ouh, that's a point. I forgot the tqdm requirement for the nasbench101 benchmark. But the paramnet data loader also uses tqdm. (line 628)

KEggensperger reviewed (changes requested):
I left some comments, but overall this looks good to me. Thanks

PhMueller committed:
[skip travis] Work on requests

PhMueller committed:
Work on requests 2

PhMueller committed:
Work on requests: Fix tests

PhMueller committed:
Fix Codestyle Error

PhMueller committed:
[skip travis] Fix#66 return by seed (#68)

* Return Sum of Runtimes. Not the mean.

* Update Nasbench101: Runindex
We can now specify a run index (0-2). The runindex is chosen in the original nasbench101 benchmark randomly.

* Reduce container size of nasbench101 container
Remove the api file from the container, instead download it before usage.
fix small bug in downloading procedure.

* Tabular Benchmark: Run Index
The runindex is set explicitly to None, then a random run id (one of 0,1,2,3) is selected.
That doesn't change the default behavior. When no run_index is specified, the run_index still is (0,1,2,3)

*  Update doc strings, Tests, Readme,

* Remove unnecessary dependencies

* Fix CodeStyle Error

PhMueller committed:
Fix Testcase

PhMueller committed:
Merge branch 'development' into ParamNet

PhMueller committed:
Trigger travis build

PhMueller committed:
Trigger build

PhMueller committed:
Fix Test

PhMueller committed:
Debug travis.

On the local machine this does not fail, but on travis.

PhMueller committed:
Fix test

PhMueller committed:
Fix test
